Automattic/newspack-ios
# Issues/58 data story folder recon

Refs #58 
The changes here are getting rather large so I'm breaking this PR into two parts.  This part will focus on the changes to add a core data layer to the Story Folders list.  The next part will focus on reconciling file system changes with what's in core data.

In this PR we're:
- Cleaning up the site folder when a user logs out.
- Refactoring the FolderStore and the FolderViewController to do its work via core data rather than dealing with file URLs directly. This effectively adds a core data layer to wrap file operations. 
- Implements an NSFetchedResultsController and a UITableViewDiffableDataSource for managing the list of folders.

Test these changes with a clean install. 

To test:
- Run unit tests, ensure everything passes. 

Scenario 1:
- Log in.  
- Check the files app to confirm you have a folder for your site.  
- Log out. 
- Check the files app and confirm that the folder was removed.

Scenario 2:
- Log in.
- View the list of story folders.
- Confirm you can create new folders.
- Confirm you can rename folders.
- Confirm you can delete folders. 
- Confirm that changes are animated.
- Tap the disclosure icon to view a detail (ignore what's in the detail view for now).
- Tap back and confirm that the site is marked as selected.
- Tap back to the menu and back to the folder list. 
- Confirm the site is still selected.

@jleandroperez game to review some fun new API in this PR?
